build.gradle

```diff
@@ -499,6 +504,32 @@ subprojects {
         maxFailures = userMaxTestRetryFailures
       }
     }
+    
+    // As we process results, check if there were any test failures.
+    afterSuite { desc, result ->
+      if (result.resultType == TestResult.ResultType.FAILURE) {
+        ext.hadFailure = true
+      }
+    }
+
+    // This closure will copy JUnit XML files out of the sub-project's build directory and into
+    // a top-level build/junit-xml directory. This is necessary to avoid reporting on tests which
+    // were not run, but instead were restored via FROM-CACHE. See KAFKA-17479 for more details.
+    doLast {
+      if (ext.isGithubActions) {
+        def dest = rootProject.layout.buildDirectory.dir("junit-xml/${project.name}").get().asFile
+        println "Copy JUnit XML for ${project.name} to $dest"
+        ant.copy(todir: "$dest") {
+          ant.fileset(dir: "${test.reports.junitXml.entryPoint}")
+        }
+
+        // If there were any test failures, we want to fail the task to prevent the failures
+        // from being cached.
+        if (ext.hadFailure) {
```

Comment:
> Personally, I'd like to make tests stable instead of retry.

Yes, agreed. 😄 Until we get the tests in better shape, I think the retry thing is okay.

 > It seems the retry issue is still existent in normal PR
 
 Normal PR will look like:
 * Any failure (flaky or not) will cause Gradle to exit with 1
 * `continue-on-error: true` will mark that step green (I think... need to verify this)
 * "Parse JUnit tests" will process the results correctly and exit 0 if only flaky, exit 1 if any true failures
 
So the overall CI workflow will be green if there are only flaky failures. Red if there are any real failures.

Reply:
> So the overall CI workflow will be green if there are only flaky failures. Red if there are any real failures.

Got it. Thanks for detailed explanation